### PR TITLE
Adds _LazyInstanceNorm and LazyInstanceNormXd

### LINKED ISSUE #60981: LazyInstanceNorm and LazyInstanceNormXd
## 🚀 Feature
Just as torch.nn has lazy BatchNorm modules, it would be nice to have lazy InstanceNorm ones as well.

## Motivation

I need `_LazyInstanceNorm` for my work on https://github.com/ToriML/onnx2pytorch (ONNX to PyTorch converter) and I've implemented them there, but it would be nice to just merge these into PyTorch for others to use as well. A lazy version of InstanceNorm is necessary for converting ONNX models because `num_features` is often known at runtime rather than conversion-to-PyTorch-time.

## Pitch

I'm about to send in a PR, so my pitch is that my forthcoming PR gets approved and merged.

## Alternatives

I could just implement LazyInstanceNorm in my own code, but I expect others to find it useful.

### PR DESCRIPTION
**Pull request by calvinmccarter-at-lightmatter**
Signed-off-by: Calvin McCarter <calvin@lightmatter.co>

Fixes #60981

### CODE REVIEW

File: torch/nn/modules/instancenorm.py [hunk 1]
```diff
@@ -59,6 +62,64 @@ def forward(self, input: Tensor) -> Tensor:
             self.training or not self.track_running_stats, self.momentum, self.eps)
 
 
+class _LazyInstanceNorm(LazyModuleMixin, _InstanceNorm):
+
+    weight: UninitializedParameter  # type: ignore[assignment]
+    bias: UninitializedParameter  # type: ignore[assignment]
+
+    cls_to_become = _InstanceNorm  # type: ignore[assignment]
```
Comment (jbschlosser) [inline torch/nn/modules/instancenorm.py:70]: nit: This isn't required since it's overwritten for the subclasses anyway

File: test/test_nn.py [hunk 2]
```diff
@@ -17627,6 +17627,119 @@ def test_lazy_batchnorm3d_state(self):
...
+                    self.assertIsInstance(lazy_module.running_var, UninitializedBuffer)
+
+                input = torch.ones(*input_shape)
+                y = lazy_module(input)
+                self.assertIsInstance(lazy_module, cls)
+                self.assertNotIsInstance(lazy_module, lazy_cls)
+
+                num_features = input_shape[1]
+                module = cls(num_features, affine=affine, track_running_stats=track_running_stats)
+                expected = module(input)
```
Comment (jbschlosser) [inline test/test_nn.py:17649]: nit: `expected` is unused - should the output of the lazy and non-lazy modules be compared?

File: test/test_nn.py [hunk 3]
```diff
@@ -17627,6 +17627,119 @@ def test_lazy_batchnorm3d_state(self):
         self._check_lazy_batchnorm_state(nn.BatchNorm3d, nn.LazyBatchNorm3d)
         self._check_lazy_batchnorm_state(nn.BatchNorm3d, nn.LazyBatchNorm3d)
 
+    def _check_lazy_instancenorm(self, cls, lazy_cls, input_shape):
```
Comment (jbschlosser) [inline test/test_nn.py:17630]: Are these `_check_lazy_instancenorm*` test functions needed? Looks like they're mostly copies of the `_check_lazy_batchnorm*` test functions, and it seems like those are already written to be fairly generic since they accept `cls` and `lazy_cls`.

I suggest repurposing `_check_lazy_batchnorm*` -> `_check_lazy_norm*` and calling those from the new lazy instance norm tests.

File: torch/nn/modules/instancenorm.py [hunk 4]
```diff
@@ -59,6 +62,64 @@ def forward(self, input: Tensor) -> Tensor:
             self.training or not self.track_running_stats, self.momentum, self.eps)
 
 
+class _LazyInstanceNorm(LazyModuleMixin, _InstanceNorm):
```
Comment (jbschlosser) [inline torch/nn/modules/instancenorm.py:65]: While this looks correct to me, there's duplication between this and `_LazyBatchNorm`. How feasible is it to alternatively implement a `_LazyNormBase` from which both `_LazyBatchNorm` and `_LazyInstanceNorm` derive?

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks for the contribution! I think the level of duplication between the lazy batch norm and lazy instance norm modules could potentially be reduced with e.g. a `_LazyNormBase`. Similarly for some of the test logic. Wdyt?

**Comment by calvinmccarter-at-lightmatter** [comment]
> Thanks for the contribution! I think the level of duplication between the lazy batch norm and lazy instance norm modules could potentially be reduced with e.g. a `_LazyNormBase`. Similarly for some of the test logic. Wdyt?

@jbschlosser -- sgtm, especially since lazy LayerNorm and GroupNorm may get added in the future. I'll get back to this PR next week.

File: torch/nn/modules/instancenorm.py (hunk 4 above)
Comment (calvinmccarter-at-lightmatter) [reply torch/nn/modules/instancenorm.py]: Done!

File: torch/nn/modules/instancenorm.py (hunk 1 above)
Comment (calvinmccarter-at-lightmatter) [reply torch/nn/modules/instancenorm.py]: Done!

File: test/test_nn.py (hunk 3 above)
Comment (calvinmccarter-at-lightmatter) [reply test/test_nn.py]: Done for `_check_lazy_norm` and `check_lazy_norm_pickle` but this is not possible for checking the state, so I've kept separate `_check_lazy_batchnorm_state` and `_check_lazy_instancenorm_state`.

File: test/test_nn.py (hunk 2 above)
Comment (calvinmccarter-at-lightmatter) [reply test/test_nn.py]: Done!

File: torch/nn/modules/batchnorm.py [hunk 5]
```diff
@@ -335,7 +336,7 @@ def _check_input_dim(self, input):
             )
 
 
-class BatchNorm2d(_BatchNorm):
+class BatchNorm2d(_LazyNormBase, _BatchNorm):
```
Comment (jbschlosser) [inline torch/nn/modules/batchnorm.py:339]: `BatchNorm2d` shouldn't be a subclass of `_LazyNormBase` - this breaks it and is causing several test failures

File: torch/nn/modules/batchnorm.py [hunk 6]
```diff
@@ -440,7 +441,7 @@ def _check_input_dim(self, input):
             raise ValueError("expected 4D input (got {}D input)".format(input.dim()))
 
 
-class BatchNorm3d(_BatchNorm):
+class BatchNorm3d(_LazyNormBase, _BatchNorm):
```
Comment (jbschlosser) [inline torch/nn/modules/batchnorm.py:444]: Similarly here, `BatchNorm3d` shouldn't be a subclass of `_LazyNormBase`

File: torch/nn/modules/batchnorm.py [hunk 7]
```diff
@@ -55,6 +55,7 @@ def __init__(
             self.register_buffer('num_batches_tracked',
                                  torch.tensor(0, dtype=torch.long,
                                               **{k: v for k, v in factory_kwargs.items() if k != 'dtype'}))
+            self.num_batches_tracked: Optional[Tensor]
```
Comment (jbschlosser) [inline torch/nn/modules/batchnorm.py:58]: With this new annotation, you'll need to make a change like `self.num_batches_tracked = 1  # type: ignore[assignment]` in `torch/distributed/pipeline/sync/batchnorm.py` to pass mypy checks.

(you can run `mypy` locally to check for errors)

**Review by jbschlosser — CHANGES_REQUESTED** [review]
Thanks for the update! Some comments below to address the failing tests

File: torch/nn/modules/batchnorm.py (hunk 7 above)
Comment (calvinmccarter-at-lightmatter) [reply torch/nn/modules/batchnorm.py:58]: I've instead added the annotation `num_batches_tracked: Tensor` in `torch/distributed/pipeline/sync/batchnorm.py`, which passes `mypy` and seems correct to me. Let me know if this is incorrect or inadvisable, though...

**Review by jbschlosser — APPROVED** [review]
LGTM! Thanks for the contribution :)
